core/drivers/versal_pm.c

```diff
@@ -0,0 +1,192 @@
+#define	PM_PINCTRL_REQUEST		28
+#define	PM_PINCTRL_RELEASE		29
+#define	PM_PINCTRL_GET_FUNCTION		30
+#define	PM_PINCTRL_SET_FUNCTION		31
+#define	PM_PINCTRL_CONFIG_PARAM_GET	32
+#define	PM_PINCTRL_CONFIG_PARAM_SET	33
+#define PM_IOCTL			34
+#define PM_QUERY_DATA			35
+#define PM_CLOCK_ENABLE			36
+#define PM_CLOCK_DISABLE		37
+#define PM_CLOCK_GETSTATE		38
+#define PM_CLOCK_SETDIVIDER		39
+#define PM_CLOCK_GETDIVIDER		40
+#define PM_CLOCK_SETRATE		41
+#define PM_CLOCK_GETRATE		42
+#define PM_CLOCK_SETPARENT		43
+#define PM_CLOCK_GETPARENT		44
+#define PM_PLL_SET_PARAMETER		48
+#define PM_PLL_GET_PARAMETER		49
+#define PM_PLL_SET_MODE			50
+#define PM_PLL_GET_MODE			51
+#define PM_FEATURE_CHECK		63
+
+/* Loader API id */
+#define PM_LOAD_PDI			0x701
+
+/* PDI sources */
+#define PDI_SRC_JTAG		0x0
+#define PDI_SRC_QSPI24		0x1
+#define PDI_SRC_QSPI32		0x2
+#define PDI_SRC_SD0		0x3
+#define PDI_SRC_EMMC0		0x4
+#define PDI_SRC_SD1		0x5
+#define PDI_SRC_EMMC1		0x6
+#define PDI_SRC_USB		0x7
+#define PDI_SRC_OSPI		0x8
+#define PDI_SRC_SBI		0x9
+#define PDI_SRC_SMAP		0xA
+#define PDI_SRC_PCIE		0xB
+#define PDI_SRC_SD1_LS		0xE
+#define PDI_SRC_DDR		0xF
+
+struct versal_sip_payload {
+	uint32_t data[PAYLOAD_ARG_CNT];
+};
+
+static uint32_t versal_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				struct versal_sip_payload *payload)
+{
+	struct thread_smc_args args = {
+		.a0 = pm_api_id > PM_LOAD_PDI ?
+		      pm_api_id : PM_SIP_SVC | pm_api_id,
+		.a1 = reg_pair_to_64(arg1, arg0),
+		.a2 = reg_pair_to_64(arg3, arg2),
+	};
+
+	thread_smccc(&args);
+
+	if (payload) {
+		reg_pair_from_64(args.a0, &payload->data[1], &payload->data[0]);
+		reg_pair_from_64(args.a1, &payload->data[3], &payload->data[2]);
+		reg_pair_from_64(args.a2, &payload->data[5], &payload->data[4]);
+	}
+
+	/* allow the PLM to output its debug information */
+	if (IS_ENABLED(CFG_VERSAL_TRACE_PLM))
+		mdelay(500);
+
+	return args.a0;
+}
+
+/* SIP call to program the FPGA has been obsoleted, use the PLM */
+TEE_Result versal_write_fpga(paddr_t pa)
+{
+	struct ipi_cmd cmd = { };
+
+	cmd.data[0] = PM_LOAD_PDI;
+	cmd.data[1] = PDI_SRC_DDR;
+	cmd.data[2] = pa >> 32;
+	cmd.data[3] = pa;
+	if (versal_mbox_notify(&cmd, NULL, NULL))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_soc_version(uint8_t *version)
+{
+	struct versal_sip_payload payload = { };
+	const uint32_t version_shift = 12;
+	uint32_t res = 0;
+
+	if (!version)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = versal_sip_call(PM_GET_CHIPID, 0, 0, 0, 0, &payload);
+	if (res) {
+		EMSG("Failed to retrieve version");
+		return TEE_ERROR_GENERIC;
+	}
+
+	*version = payload.data[2] >> version_shift;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result versal_check_pm_abi(void)
+{
+	struct versal_sip_payload p = { };
+	struct ipi_cmd cmd = { };
+	struct ipi_cmd rsp = { };
+
+	/* Check with TF-A */
+	if (versal_sip_call(VERSAL_SIP_SVC_VERSION, 0, 0, 0, 0, &p))
+		panic();
```

Comment:
Is panic necessary, can't we just complain a bit and avoid using this?
By the way, I guess we should check `VERSAL_SIP_SVC_UID` to see that it really is a Versal SIP we're talking to and not some other SIP.

Reply:
Yes for our PM usage panic is not necessary - maybe for some of the other services, will just report the error.

Sure I can use the UID. BTW should the checks against the values be with CFG_* instead of built-in macros? Just to avoid having to modify code for updates

Reply:
No, it's enough to check that the major and minor versions are larger or equal to the expected value. This is explained in the SMC Calling Convention specification.

I don't expect the SIP UID to change.

Reply:
ok. I followed what was done in `optee_msg_api_revision_is_compatible` that you pointed to below (so a match on expected major and minor >= expected minor)